interfaces/builtin/service_watchdog.go

```diff
@@ -50,18 +50,41 @@ type serviceWatchdogInterface struct {
 var osGetenv = os.Getenv
 
 func (iface *serviceWatchdogInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	// If the system has defined it, use NOTIFY_SOCKET from the environment. Note
+	// this is safe because it is examined on snapd start and snaps cannot manipulate
+	// the environment of snapd.
 	notifySocket := osGetenv("NOTIFY_SOCKET")
 	if notifySocket == "" {
 		notifySocket = "/run/systemd/notify"
-	} else {
-		// must be an absolute path without any AppArmor regular
-		// expression (AARE) characters or double quotes
-		if !strings.HasPrefix(notifySocket, "/") || strings.ContainsAny(notifySocket, `?*[]{}^"`) {
-			return fmt.Errorf("cannot use %q as notify socket path", notifySocket)
-		}
 	}
+
+	if !(strings.HasPrefix(notifySocket, "/") || strings.HasPrefix(notifySocket, "@")) || strings.ContainsAny(notifySocket, `?*[]{}^"`) {
```

Comment:
Wait, the logic is wrong here. if notifySocket is "[" then this will not error out.

Reply:
Double checked, the code returns the error `cannot use "[" as notify socket path` as expected. Note that the check is equivalent to: `!(prefix is either '/' or '@') or socket contains any of ?*[]{}^`